### PR TITLE
fix 'BG data unchanged' check

### PR DESCRIPTION
**Pull request by justmara**
fix for #1741

The stale bg data check moved to SafetyCheck constraint making in plugin-independent. Also [fixed logic](https://github.com/nightscout/AndroidAPS/issues/1741#issuecomment-1295771666) of determining flat (stale) BG data.

Original (erroneous) checks removed from determine-basal.

### CODE REVIEW

File: app/src/main/assets/AIMI/variant Pam/determine-basal.js [hunk 1]
```diff
@@ -0,0 +1,1885 @@
+/*
```
Comment (jbr7rr) [inline app/src/main/assets/AIMI/variant Pam/determine-basal.js:1]: Oops, these files are not supposed to be here right?

File: core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt [hunk 2]
```diff
@@ -100,8 +102,35 @@ class GlucoseStatusProvider @Inject constructor(
         ).also { aapsLogger.debug(LTag.GLUCOSE, it.log()) }.asRounded()
     }
 
-    companion object {
+    fun isBgFlatForInterval(minutes: Long, maxDelta: Double) : Boolean? {
```
Comment (jbr7rr) [inline core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: Missing the original isSaveCgmSource check here which is set to check if the source is Dexcom (BYODA)

File: app/src/main/assets/AIMI/variant Pam/determine-basal.js (hunk 1 above)
Comment (justmara) [reply app/src/main/assets/AIMI/variant Pam/determine-basal.js:1]: woah, sorry :) pushing without it. took commit from my branch

File: core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt (hunk 2 above)
Comment (justmara) [reply core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: I didn't changed that check at all. As I've said it _can_ be changed to respect this smoothing option too, but imho that must be an additional change, subject for further discussion.

Comment (jbr7rr) [reply core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: Well in the old behavoir (in dev) the Dexcom plugin is excluded from the 45min unchanged check, this is not included in your PR as far as I can see. 

Current implementation in dev:

variable saveCgmSource is set here:
https://github.com/nightscout/AndroidAPS/blob/b4edb71e279954a1aa93a9ac720898d57494221e/app/src/main/java/info/nightscout/androidaps/plugins/aps/openAPSSMB/OpenAPSSMBPlugin.kt#L210

And is used in determine_basal.js here
https://github.com/nightscout/AndroidAPS/blob/b4edb71e279954a1aa93a9ac720898d57494221e/app/src/main/assets/OpenAPSSMB/determine-basal.js#L113

Comment (justmara) [reply core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: ahh, this one. ok, i'll re-introduce it here too, ok

Comment (justmara) [reply core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: pushed update

Comment (justmara) [reply core/src/main/java/info/nightscout/androidaps/plugins/iob/iobCobCalculator/GlucoseStatusProvider.kt:105]: the variable `isSaveCgmSource` is still used in determine-basal for some conditions so seems that it must be reviewed and replaced with something like `byoda || smoothed`, but there are still some points to discuss:

1. byoda with g7 does not have smoothing currently, so the check for byoda becomes not so trustful
2. this check must be moved somewhere else to be shared and not copy-pasted by plugins.

**Comment by ga-zelle** [comment]
@justmara I read the arlier attempts to get this test improved. The final problem in the past was lack of buy-in from US/OpenAPS. Therefore I contacted Scott Lybrand and discussed the matter with him. He has no objection to using a better check for flatness than the one based on the 3 deltas. As a result I suggested a check similar to yours, but slightly different in 3 aspetcs.

Here is the code snippet I inserted in GlucodeStatusProvider.kt right after where the 3 deltas are determined:
```
// calculate length of cgm values staying within a narrow band
val bandWidth = 2                   // max allowed width; how about mmol/L conversions?
var minBG = now.value
var maxBG = minBG
var cgmFlatMinutes = 0.0
var oldDate = now.timestamp
for (i in 1 until sizeRecords) {
    val then = data[i]
    val thenDate = then.timestamp
    minBG = Math.min(minBG, then.value)
    maxBG = Math.max(maxBG, then.value)
    //          outside band            or pause > 11 minutes       or older than 1 hour
    if (maxBG-minBG > bandWidth || oldDate-thenDate > 1000*60*11 || nowDate-thenDate > 1000*60*60) {
        break
    } else {
        oldDate = thenDate
    }
cgmFlatMinutes = ( nowDate-oldDate ) / 60000.0
}
```
The first difference is a bandwidth of 2mg/dl for flatness because a CGM working in mmol/l is allowed a difference of 0.1mmol/l which is converted internally to 2mg/dl.

The second difference is using min and max values for the time window and ensuring that difference is not more than 2mg/dl. Your current check depends on whether the initial value happens to be in the middle if that band or on the edge.

The third diffference is that instead of true/false for fixed 45minutes I return the length of that time window . The orginal 45m came from a "quick fix" at the time because the 3 deltas were available. Now a more meaningful length can be picked like 30m inside determine-basal to switch to fail-safe mode and another length like 60m for raising any alarm like "check your CGM".

**Comment by MilosKozak** [comment]
I did smoother integration in https://github.com/nightscout/AndroidAPS/commit/d5af299692efb0e05762dc0ccf1f38f49836cbdf
hopefully I fixed all bugs from your implementation
comments welcome

**Pull request closed without merging**